File: mg01/game/tech/Assets/Resources/LuaScript/src/org/ref/Refer.lua
--Refer
--@author jr.zeng
--2017年10月11日 下午12:00:47
local modname = "Refer"
--==================global reference=======================

local type = type
local string = string

--===================namespace========================
local ns = "org"
local using = {--[[namespace1,...]]}
local _ENV = namespace(ns, using, modname)
--===================module property========================
local super = nil    --父类
Refer = {}
local Refer = Refer

--===================module content========================

--@must_ 必须传refer
function Refer.format(refer_, must_)

	if refer_ == nil then
		if must_ ~= false then
			Refer.assert(refer_)
		end
		return nil
	end

	local str
	if type(refer_) == "string" then
		str = refer_
	elseif refer_.referId then
		str = refer_.referId
	else
		log.error("不支持的Refer", refer_)
	end

	if string.isNullOrEmpty(str) then
		return nil
	end

	return str
end

function Refer.assert(refer_, error)

	if not refer_ or not refer_.referId then
		log.error(error or "此对象不是Refer ", refer_)
		return false
	end
	return true
end

function Refer.dump()

end

-------∽-★-∽------∽-★-∽Notify∽-★-∽------∽-★-∽--------//

local __subDeactive
local __subDispose

function Refer.a()  --TODO
	__subDeactive = new(Subject)
	__subDispose =  new(Subject)
end

function Refer.clearNotify()
	__subDeactive:detachAll()
	__subDispose:detachAll()
end

-- 通知沉默
function Refer.notifyDeactive(refer_)
	local referId = Refer.format(refer_)
	--nslog.print_t('notifyDeactive', referId)
	__subDeactive:notify(referId, referId)
end

function Refer.attachDeactive(refer_, fun, target)
	local referId = Refer.format(refer_)
	--nslog.print_t("attachDeactive", referId)
	__subDeactive:attach(referId, fun, target, false)    --这里不能传入refer，要主动detachDeactive
end

function Refer.detachDeactive(refer_, fun, target)
	local referId = Refer.format(refer_)
	__subDeactive:detach(referId, fun, target)
end

-- 通知析构
function Refer.notifyDispose(refer_)
	local referId = Refer.format(refer_)
	__subDispose:notify(referId, referId)
end

function Refer.attachDispose(refer_, fun, target)
	local referId = Refer.format(refer_)
	__subDispose:attach(referId, fun, target, false)    --这里不能传入refer，要主动detachDispose
end

function Refer.detachDispose(refer_, fun, target)
	local referId = Refer.format(refer_)
	__subDispose:detach(referId, fun, target)
end


return Refer

File: mg01/game/tech/Assets/Script/Org/Ref/Refer.cs
/* ==============================================================================
 * Refer
 * @author jr.zeng
 * 2017/6/4 14:31:17
 * ==============================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


using UnityEngine;
using Object = UnityEngine.Object;

namespace mg.org
{


    //目前用于：
    //1. ImageAbs解决go托管问题
    //2. AssetCache引用计数
    //3. Subject不用手动Detach
        

    public class Refer //: IRefer   //不做基类使用了
    {
        //string m_referId;

        //public Refer(string referId_)
        //{
        //    m_referId = referId_;
        //}

        ///// <summary>
        ///// 引用者id
        ///// </summary>
        //public string ReferId
        //{
        //    get { return m_referId; }
        //}

        //public void NotifyDeactive()
        //{
        //    NotifyDeactive(this);
        //}
        //public void NotifyDispose()
        //{
        //    NotifyDispose(this);
        //}


        //-------∽-★-∽------∽-★-∽--------∽-★-∽static∽-★-∽--------∽-★-∽------∽-★-∽--------//

        /// <summary>
        /// 获取引用id
        /// </summary>
        /// <param name="refer_"></param>
        /// <returns></returns>
        static public string Format(object refer_)
        {
            if (refer_ == null)
                return null;
            
            string str;
            if (refer_ is string)
            {
                str = refer_ as string;
            }
            else if (refer_ is IRefer)
            {
                str = (refer_ as IRefer).ReferId;
            }
            else
            {
                Log.Assert("不支持的Refer: " + refer_);
                return null;
            }

            if (string.IsNullOrEmpty(str))
                return null;

            return str;
        }

        static public bool Assert(object refer_)
        {
            if (refer_ is string)
            {
                return true;
            }
            else if (refer_ is IRefer)
            {
                return true;
            }

            Log.Assert("不正确的Refer: " + refer_);
            return false;
        }


        /// <summary>
        /// 转换为引用id
        /// </summary>
        /// <param name="refer_"></param>
        /// <returns></returns>
        static public string TransID(UnityEngine.Object refer_)
        {
            return refer_.GetType().Name + "#U3D#" + refer_.GetInstanceID();
        }


        static public void Dump()
        {
            //TODO
        }

        //static ISubject __subDeactive = new Subject();
        static ISubject __subDeactive = new Notifer();
        //static ISubject __subDispose = new Subject();
        static ISubject __subDispose = new Notifer();

        static public void ClearNotify()
        {
            __subDeactive.DetachAll();
            __subDispose.DetachAll();
        }

        //-------∽-★-∽------∽-★-∽NotifyDeactive∽-★-∽------∽-★-∽--------//

        /// <summary>
        /// 通知沉默
        /// </summary>
        /// <param name="refer_"></param>
        static public void NotifyDeactive(object refer_)
        {
            string referId = Format(refer_);
            if( __subDeactive.Notify(referId, referId) )
            {
                //通知完移除全部监听
                __subDeactive.DetachByType(referId);
            }

            __subDeactive.Notify(GL_EVENT.REFER_DEACTIVE, refer_);
        }

        static public void AttachDeactive(object refer_, CALLBACK_1 callback_)
        {
            string referId = Format(refer_);
            __subDeactive.Attach(referId, callback_, null);
        }

        static public void DetachDeactive(object refer_, CALLBACK_1 callback_)
        {
            string referId = Format(refer_);
            __subDeactive.Detach(referId, callback_);
        }

        static public void AttachDeactive(CALLBACK_1 callback_)
        {
            __subDeactive.Attach(GL_EVENT.REFER_DEACTIVE, callback_, null);
        }
        static public void DetachDeactive(CALLBACK_1 callback_)
        {
            __subDeactive.Detach(GL_EVENT.REFER_DEACTIVE, callback_);
        }

        //-------∽-★-∽------∽-★-∽NotifyDispose∽-★-∽------∽-★-∽--------//

        /// <summary>
        /// 通知析构
        /// </summary>
        /// <param name="refer_"></param>
        static public void NotifyDispose(object refer_)
        {
            string referId = Format(refer_);
            if (__subDispose.Notify(referId, referId) )
            {
                //通知完移除全部监听
                __subDispose.DetachByType(referId);
            }

            //__subDeactive.Notify(GL_EVENT.REFER_DISPOSE, refer_);
        }

        /// <summary>
        /// 监听析构，(暂时是retain用到)
        /// </summary>
        /// <param name="refer_"></param>
        /// <param name="callback_"></param>
        static public void AttachDispose(object refer_, CALLBACK_1 callback_)
        {
            string referId = Format(refer_);
            __subDispose.Attach(referId, callback_, null);
        }

        static public void DetachDispose(object refer_, CALLBACK_1 callback_)
        {
            string referId = Format(refer_);
            __subDispose.Detach(referId, callback_);
        }

        static public void AttachDispose(CALLBACK_1 callback_)
        {
            __subDeactive.Attach(GL_EVENT.REFER_DISPOSE, callback_, null);
        }

        static public void DetachDispose(CALLBACK_1 callback_)
        {
            __subDeactive.Detach(GL_EVENT.REFER_DISPOSE, callback_);
        }


    }
}